include/arch/arm/arch/64/mode/machine.h

```diff
@@ -380,5 +380,18 @@ void arch_clean_invalidate_L1_caches(word_t type);
 
 static inline paddr_t addressTranslateS1(vptr_t vaddr)
 {
-    return ats1e1r(vaddr);
+    word_t result;
+    UNUSED word_t saved_par;
+    if (config_set(CONFIG_ARM_HYPERVISOR_SUPPORT)) {
+        MRS("par_el1", saved_par);
+        isb();
+    }
+
+    result = ats1e1r(vaddr);
+
+    if (config_set(CONFIG_ARM_HYPERVISOR_SUPPORT)) {
+        MSR("par_el1", saved_par);
+    }
```

Comment:
```suggestion
    if (config_set(CONFIG_ARM_HYPERVISOR_SUPPORT)) {
        /* Prevent re-ordering with the read of PAR above */
        isb();
        MSR("par_el1", saved_par);
    }
```

Reply:
Added, however, still not clear if there is already a dependecy read then write it is similar to read modify write approach.

Reply:
> Added, however, still not clear if there is already a dependecy read then write it is similar to read modify write approach.

The cpu only recognises such dependencies with normal registers, it doesn't know that AT affects PAR, otherwise we wouldn't need any ISBs around AT.

Whether it reorders system register writes and reads to the same system register is another matter, perhaps it's guaranteed not to do that, then we can drop one ISB. However, it does say you need an ISB after a system reg write in general, with a few exceptions, which implies there are few guarantees.

Edit: Moved to here and expanded reply.

If you do a read, modify write operation on the same system reg a dependency is introduced via a normal register, so no ISB or anything is needed. E.g. `MRS x1 N`, `MSR N x1`.

On pipelined CPUs, to prevent a pipeline flush and having to wait the full e.g. 20 cycles till the first operation is finished before starting the dependent instruction, CPUs have a bypass network which forwards results quicker to the next, dependent instruction so there only need to be a few cycles delay before executing the dependent instruction. The next instruction gets its input register values from the bypass network instead of the register file. 

With out-of-order execution the CPU keeps track of all memory read/write dependencies needed for consistency in addition to all register dependencies, and that's where memory ordering comes in. But even with in-order execution, instructions could finish out-of-order because some instructions take longer than others, and this could be an issue with a MSR after an MRS.

All that to say that for normal instructions things are well-defined because of natural register dependencies and for memory accesses they take great care to define the model in the documentation, but for system registers things are fuzzy and I haven't found anything except the requirement for ISB after most MSR if you want the effect to be visible.

MRS reads can be re-ordered safely with each other because they don't have side-effects, but how are MSR treated in regard to MRS? Are they just executed independently and if there is an ordering requirement you have to either force it via normal registers or ISB? Or are they serialised and executed in-order in regard to each-other? Conceptually this would be treating all system register as one normal register.